Documenting HBI identity logic (#900)
RedHatInsights/insights-host-inventory: docs/_index.md

In addition to the /hosts/{host_id_list}/system_profile endpoint, this query can also be used on the upper-level /hosts endpoint to fetch partial system profile data for hosts. On the /hosts endpoint, if this query is not included, no system profile data will be returned.

## Identity Requirements

To create, update and access host data, the provided identity must comply with the following logic:
- it must contain an account number
- identity must be of type `System` or `User`
- the auth type in identity must of one of:
    - `basic_auth`, `cert-auth`, `classic-proxy`, `jwt-auth` or `uhc-auth`

> If the identity is of type `User`, there are no additional requirements. However, if the identity type is `System`, the `system` object should be present in identity and the following logic will be used to validate:

Reviewer: Needs update to describe when `identity.type == "System"` and `identity.auth_type == "classic-proxy"`.   See https://github.com/RedHatInsights/insights-host-inventory/pull/902

Author: @thearifismail I think you meant auth type is `classic-proxy`, there are no additional requirements.

In addition to the /hosts/{host_id_list}/system_profile endpoint, this query can also be used on the upper-level /hosts endpoint to fetch partial system profile data for hosts. On the /hosts endpoint, if this query is not included, no system profile data will be returned.

## Identity Requirements

> To create, update and access host data, the identity header must comply with the following logic.

Reviewer 2: ```suggestion
To create, update and access host data, the provided identity must comply with the following logic:
```
I'm only removing "header" because the Identity can be provided via header or platform_metadata

- it must contain an account number
- identity must be of type `System` or `User`
- the auth type in identity must of one of:
    - `basic_auth`, `cert-auth`, `classic-auth`, `jwt-auth` or `uhc-auth`

> If the identity is of type `User`, there is not any additional requirements. However, ff the identity type is `System`, the `system` object should be present in identity and the following identity logic will be used to validate identity.

Reviewer 2: ```suggestion
If the identity is of type `User`, there are no additional requirements. However, if the identity type is `System`, the `system` object should be present in identity and the following logic will be used to validate:
```

- the fields `cert-type` and `cn` must be present within the `system` object
- the `cert-type` must be one of:
    - `hypervisor`, `rhui`, `sam`, `satellite` or `system`
- the value provided in `cn` must be equal to the `owner_id` of the host being added or updated

> Host update messages received over `platform.inventory.system-profile` topic does not need to provide an identity header.

Reviewer 2: ```suggestion
Host update messages received over `platform.inventory.system-profile` topic do not need to provide an identity.
```

> The following must be true for host add or update messages received over `platform.inventory.events`.

Reviewer 2: ```suggestion
The following must be true for `add_host` messages received over `platform.inventory.host-ingress` and `platform.inventory.host-ingress-p1`.
```

> - they must contain `platform_metadata` and the `identity` field must be encoded within the `platform_metadata`, except when the reporter is `rhsm-conduit`

Reviewer 2: ```suggestion
- they must contain `platform_metadata` and the `b64_identity` field must be encoded within the `platform_metadata`, except when the reporter is `rhsm-conduit`
```

> - providing identity is optional when the reporter is `rhsm-conduit`

Reviewer 2: I think this one might be redundant after the previous bullet point

> - the `owner_id` of the host must match the value in `system.cn` of the identity.

Reviewer 2: I just wanna clarify that this requirement is only for updating existing hosts; if no match is found, a new host is created

Author: just added a clause to this point :+1:

> If the reporter is `rhsm-conduit`, and an identity is not provided, adding or updating hosts over the `platform.inventory.events` will follow the logic below.

Reviewer 2: ```suggestion
If the reporter is `rhsm-conduit` and an identity is not provided, adding or updating hosts over `platform.inventory.host-ingress` and `platform.inventory.host-ingress-p1` will follow this logic:
```

> - a `subscription_manager_id` must be present

Reviewer 2: ```suggestion
- `subscription_manager_id` must be present
```

> - `subscription_manager_id` of the hosts gets set as its `owner_id`

Reviewer 2: ```suggestion
- the host's `owner_id` gets set to its `subscription_manager_id`
```

> - if `owner_id` of a host does not equate its `subscription_manager_id`, it gets overwritten to the value of `subscription_manager_id`

Reviewer 2: I think this one is redundant, since its `owner_id` gets set to its `subscription_manager_id` regardless

## Identity Requirements

To create, update and access host data, the provided identity must comply with the following logic:
- it must contain an account number
- identity must be of type `System` or `User`
- the auth type in identity must of one of:

>     - `basic_auth`, `cert-auth`, `classic-auth`, `jwt-auth` or `uhc-auth`

Reviewer: `classic-auth` should be `classic-proxy`, https://github.com/RedHatInsights/insights-host-inventory/blob/f33d27d9e3fc3d5426fbc56f9f2c9b281c9b4aab/app/auth/identity.py#L30

If the identity is of type `User`, there are no additional requirements. However, if the identity type is `System`, the `system` object should be present in identity and the following logic will be used to validate:
- the fields `cert-type` and `cn` must be present within the `system` object
- the `cert-type` must be one of:
    - `hypervisor`, `rhui`, `sam`, `satellite` or `system`

> - the value provided in `cn` must be equal to the `owner_id` of the host being added or updated

Reviewer: the `owner_id` in the host.system_profile being added or update.
